### PR TITLE
Atmosphere Incorporation: USatm1976, MIL_SPEC_210A Hot/Cold/Polar/Tropical, and Offsets

### LINKED ISSUE #133: Add ability to use hot day (non standard atmosphere)
### Summary of Issue

GASP had the ability to include a temperature delta in its mission analysis for a hot day, which was especially useful for acoustic missions. We should add this ability to Aviary.

### Issue Type

- [ ] Bug
- [x] Enhancement
- [ ] Code Cleanup
- [ ] Docs
- [ ] Miscellaneous

### Is your feature request related to a problem? Please describe.

_No response_

### Associated Bug Report

_No response_

### PR DESCRIPTION
**Pull request by ehariton**
### Summary

Incorporates USATM1976 directly into aviary as the default atmosphere so we no longer need to rely on the dymos atmo library. 

Incorporates open-source MIL_SPEC_210A atmospheres (hot/cold/polar/tropical) which the user can switch to via options. 

Individual atmosphere data is now separated into it's own file, while the core component that imports the data and does the calculations has been named `atmosphere.py`. 

Adds an temperature offset option, allowing the user to shift an atmosphere profile by a specific number of degrees kelvin to mimic hotter / colder days.

Geometric vs. Geopotential input option is preserved.

Removed option to add `dsos_dh` as an output. This is now always provided. This is only used by solved 2DOF ODE.

Removed `drhos_dh` output as it was not used in Aviary.

Many tests that relied on atmosphere, including some engine conversion tests, referenced the atmosphere model. Because the original dymos atmosphere model was based on English units and used an English akima spline, the values changed slightly when converted to the  SI units and the new akima splines. SI units is what the original 1976 atm standard was published in. There were noticeable small differences, rounding etc, between the dymos atmosphere implementation in imperial units and the values published in SI units in the original standard.

### Related Issues

- Resolves #133 

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 1]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+        elif self.options['data_source'] == 'tropical':
+            self.source_data = tropical_210A
+        elif self.options['data_source'] == 'polar':
+            self.source_data = polar_210A
+        elif self.options['data_source'] == 'hot':
+            self.source_data = hot_210A
+        elif self.options['data_source'] == 'cold':
+            self.source_data = cold_210A
+        else:
+            Warning(
```
Comment (Kenneth-T-Moore) [inline aviary/subsystems/atmosphere/atmosphere.py]: This warning isn't needed (and is impossible to reach), because the OpenMDAO options dictionary typechecks this for us.

File: aviary/subsystems/atmosphere/test/test_atmosphere.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
+import unittest
+
+import numpy as np
+import openmdao.api as om
+from openmdao.utils.assert_utils import assert_check_partials, assert_near_equal
+
+from aviary.subsystems.atmosphere.atmosphere import AtmosphereComp
+
+
+class USatm1976TestCase1(unittest.TestCase):
```
Comment (Kenneth-T-Moore) [inline aviary/subsystems/atmosphere/test/test_atmosphere.py:56]: Should add the geodetic/geopotential tests from here: https://github.com/OpenMDAO/dymos/blob/master/dymos/models/atmosphere/test/test_atmos.py

Also, I think a test would be useful where you put both the OpenMDAO and dymos atmospehere in an empty openmndao model and compare that the answers are the same.

File: aviary/docs/user_guide/atmosphere.ipynb
Comment (jkirk5) [inline aviary/docs/user_guide/atmosphere.ipynb:1]: Can you add a mention of the actual names of the standards Aviary uses for its atmospheres, and what they do? I.E. 1976 standard for a general-purpose average atmosphere + the milspec for various climates

File: aviary/models/engines/turbofan_23k_1.csv [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# created 06/06/25
+# created 01/08/26 at 14:09 by omdao
```
Comment (jkirk5) [inline aviary/models/engines/turbofan_23k_1.csv]:
```suggestion
# created 01/08/26
```

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 4]
```diff
@@ -26,29 +33,33 @@ def initialize(self):
...
             'input_speed_type',
             default=SpeedType.TAS,
             types=SpeedType,
             desc='defines input airspeed as equivalent airspeed, true airspeed, or mach number',
         )
 
+        self.options.declare(
+            'delta_T_Kelvin',
+            default=0.0,
+            desc='Temperature delta from International Standard Atmosphere (ISA) standard day conditions (degrees Kelvine)',
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]:
```suggestion
            desc='Temperature delta from International Standard Atmosphere (ISA) standard day conditions (degrees Kelvin)',
```

Comment (jkirk5) [reply aviary/subsystems/atmosphere/atmosphere.py]: Deg C and K have the same magnitude so it might be more natural language to state that the delta is in Celsius

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 5]
```diff
@@ -26,29 +33,33 @@ def initialize(self):
...
+        )
+
+        self.options.declare(
+            'data_source',
+            default='USatm1976',
+            desc='The atmospheric model used. Chose one of USatm1976, tropical, polar, hot, cold.',
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: Can we simply call it "standard"? It is much simpler to remember than a very specific shortening of "1967 standard atmosphere", and we are using generic terms for the milspec atmospheres anyway

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 6]
```diff
@@ -26,29 +33,33 @@ def initialize(self):
...
             desc='defines input airspeed as equivalent airspeed, true airspeed, or mach number',
         )
 
+        self.options.declare(
+            'delta_T_Kelvin',
+            default=0.0,
+            desc='Temperature delta from International Standard Atmosphere (ISA) standard day conditions (degrees Kelvine)',
+        )
+
+        self.options.declare(
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: This should get updated to use a new Enum that is linked to a new Aviary variable, so users can set this from the csv. Maybe in `Mission`?. Adding a `types` argument here pointing to that Enum will also catch incorrect options getting passed - see the speed type option for an example

Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: Also, this option never gets used in this component, so I'm pretty sure it can get removed and then only exist in `AtmosphereComp`. I'm pretty sure `set_model_options()` handles getting it where it needs to go if it is an Aviary variable

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 7]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+        ----------
+        inputs : `Vector`
+            `Vector` containing inputs.
+        outputs : `Vector`
+            `Vector` containing outputs.
+        """
+        table_points = self.source_data.alt
+        h = inputs['h']
+
+        if self._geodetic:
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py:178]: FLOPS had a built-in equation for conversion (it exists in `subsystems/propulsion/utils.py`, `convert_geopotential_altitude`). It gets used to convert engine decks altitudes if needed, should these get combined?

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 8]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+    def setup(self):
+        """
+        Add component inputs and outputs.
+        """
+        nn = self.options['num_nodes']
+
+        self._dt = self.options['delta_T_Kelvin']
+
+        self._geodetic = self.options['h_def'] == 'geodetic'
+        self._R0 = 6_356_766  # (meters) The effective Earth Radius
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py:119]: add to `aviary/constants.py`!

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 9]
```diff
@@ -26,29 +33,33 @@ def initialize(self):
...
         nn = self.options['num_nodes']
         speed_type = self.options['input_speed_type']
         h_def = self.options['h_def']
-        output_dsos_dh = self.options['output_dsos_dh']
 
         self.add_subsystem(
             name='standard_atmosphere',
-            subsys=USatm1976Comp(num_nodes=nn, h_def=h_def, output_dsos_dh=output_dsos_dh),
+            subsys=AtmosphereComp(num_nodes=nn, h_def=h_def),
             promotes_inputs=[('h', Dynamic.Mission.ALTITUDE)],
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py:59]: Since this is a built-in Aviary component now we shouldn't need aliasing anymore - our atmosphere components can directly output the correct Aviary variables

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 10]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+        coeffs = self.source_data.akima_P[idx]
+        outputs['pres'] = pressure = coeffs[:, 0] + dx * (
+            coeffs[:, 1] + dx * (coeffs[:, 2] + dx * coeffs[:, 3])
+        )
+
+        coeffs = self.source_data.akima_rho[idx]
+        raw_density = coeffs[:, 0] + dx * (coeffs[:, 1] + dx * (coeffs[:, 2] + dx * coeffs[:, 3]))
+
+        # Equation 42, rho = (P * M)/(R * (T + dT))
+        # Assumes pressure does not change (which is a simplification)
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py:203]: Thank you for detailed comments on the method! This will probably be very helpful someday to figure out what the component is doing

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 11]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+    if units == 'SI':
+        atm_data.P *= 100  # mb -> pascal
+    elif units == 'English':
+        atm_data.alt *= 0.3048  # ft -> m
+        atm_data.T = (atm_data.T - 32) * 5 / 9 + 273.15  # degF -> degK
+        atm_data.P *= 3376.85  # inHg60 -> Pascal
+        atm_data.rho *= 0.453592 / (0.3048**3)  # lbm/ft**3 -> kg/m**3
+    else:
+        print(f"units must be SI or English but '{units}' was supplied.")
+        exit()
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: No bare print statements please, put behind a verbosity check and make it into a warning, don't use `exit`!

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 12]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+    # Units have now been translated into (altitude (meters), temperature (degK), pressure(pascals), dynamic viscosity (kg/m**3))
+
+    import textwrap
+    from openmdao.components.interp_util.interp import InterpND
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: Is there a reason to have imports down here, and not at the top with the others?

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 13]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+            partials['pres', 'h'][...] *= dz_dh
+            partials['rho', 'h'][...] *= dz_dh
+            partials['viscosity', 'h'][...] *= dz_dh
+            partials['sos', 'h'][...] *= dz_dh
+            partials['dsos_dh', 'h'] *= dz_dh**2
+
+
+def _build_akima_coefs(out_stream, raw_data, units):
+    """
+    Print out the Akima coefficients based on the raw atmospheric data.
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: Print is an optional feature of this function, change to "compute" or "generate"?

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 14]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+        # print('Pressure (Pa)', prob.get_val('pres', units='Pa'))
+        # print('Density (kg/m**3)', prob.get_val('rho', units='kg/m**3'))
+        # print('Viscosity (Pa*s)', prob.get_val('viscosity', units='Pa*s'))
+        # print('Speed of Sound (m/s)', prob.get_val('sos', units='m/s'))
+
+        print('Temperatures (degF):', prob.get_val('temp', units='degF'))
+        print('Pressure (inHg60)', prob.get_val('pres', units='inHg60'))
+        print('Density (lbm/ft**3)', prob.get_val('rho', units='lbm/ft**3'))
+        print('Viscosity (Pa*s)', prob.get_val('viscosity', units='Pa*s'))
+        print('Speed of Sound (m/s)', prob.get_val('sos', units='m/s'))
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: I don't like making this file a script - no other Aviary files except the examples & tests do that, so it is very unexpected behavior for this file. It should probably get moved the same place as the actual akima builder per my other comment on that function

File: aviary/docs/user_guide/atmosphere.ipynb [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+    "\n",
+    "## Create a new Atmosphere file\n",
+    "You will need to make a new file to hold the raw data of your atmosphere model and the akima splines. This file should be titled after your atmosphere model, i.e. `StandardAtm1976` or `MIL_SPEC_210A_Cold`. \n",
+    "Save that file inside of the `aviary/subsystems/atmosphere` folder.\n",
+    "You will need to make a new `_raw_data` dictionary inside your new file. Look at the example `_raw_data` contained in `StandardAtm1976` for reference. \n",
+    "\n",
+    "## Converting Raw data Proper Units\n",
+    "From here on our we need to be especially mindful of units. We are going to use a non-openmdao helper function `_build_akima_coefs` to help create the akima coefficients. \n",
```
Comment (jkirk5) [inline aviary/docs/user_guide/atmosphere.ipynb]:
```suggestion
    "From here on our we need to be especially mindful of units. We are going to use a helper function `_build_akima_coefs` to help create the akima coefficients. \n",
```

I don't think specifying non-openmdao adds any useful info. It's also not a dymos, numpy, pyoptsparse, etc. helper function so why single out OM?

File: aviary/subsystems/atmosphere/atmosphere.py [hunk 16]
```diff
@@ -64,3 +75,420 @@ def setup(self):
...
+        if self._geodetic:
+            partials['temp', 'h'][...] *= dz_dh
+            partials['pres', 'h'][...] *= dz_dh
+            partials['rho', 'h'][...] *= dz_dh
+            partials['viscosity', 'h'][...] *= dz_dh
+            partials['sos', 'h'][...] *= dz_dh
+            partials['dsos_dh', 'h'] *= dz_dh**2
+
+
+def _build_akima_coefs(out_stream, raw_data, units):
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/atmosphere.py]: This should probably get moved to a util folder, it feels odd to have here. Maybe an `atmosphere/utils.py` file?

File: aviary/subsystems/atmosphere/data/MIL_SPEC_210A_Cold.py
Comment (jkirk5) [inline aviary/subsystems/atmosphere/data/MIL_SPEC_210A_Cold.py:1]: Can you move all the atmosphere data files to an `atmosphere/data` folder?

File: aviary/docs/user_guide/atmosphere.ipynb [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+    "from aviary.subsystems.atmosphere.MIL_SPEC_210A_Tropical import DATA_ORIGIN_NOTE\n",
+    "\n",
+    "print(DATA_ORIGIN_NOTE)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Adding your own Atmosphere Model\n",
```
Comment (jkirk5) [inline aviary/docs/user_guide/atmosphere.ipynb:115]: Maybe this is developer guide atmosphere info. I'm starting to think we need to completely re-organize how we organize the docs anyway though, splitting up the subsystems into different places doesn't seem useful...

File: aviary/subsystems/atmosphere/data/MIL_SPEC_210A_Hot.py [hunk 18]
```diff
@@ -0,0 +1,1119 @@
...
+        Geopotential Altitude in meters.
+    temp : float
+        Temperature in degK.
+    pres : float
+        Pressure in pascal.
+    rho : float
+        Density in kg/m^3.
+    """
+
+# turn off ruff linting
```
Comment (jkirk5) [inline aviary/subsystems/atmosphere/data/MIL_SPEC_210A_Hot.py]: Technically formatting, linting is when actual code besides whitespace gets touched

File: aviary/subsystems/atmosphere/atmosphere.py (hunk 16 above)
Comment (jkirk5) [reply aviary/subsystems/atmosphere/atmosphere.py]: I also just noticed it probably shouldn't be a private function if we want it to be used by users

File: aviary/docs/user_guide/atmosphere.ipynb
Comment (ehariton) [reply aviary/docs/user_guide/atmosphere.ipynb:1]: covered by a few sentences in the first paragraph:

"The standard 1976 tables are the default option and represent the typical conditions experienced at a given altitude. However, it is important to note that "the atmosphere has never, and will never be, standard". All of the atmosphere models are approximations. The MIL-SPEC-210A tables represent extreme atmospheres in tropical, arctic (polar winter), extreme hot, and extreme cold climates. "

File: aviary/subsystems/atmosphere/atmosphere.py (hunk 5 above)
Comment (ehariton) [reply aviary/subsystems/atmosphere/atmosphere.py]: Sounds good to me.
